spark_expectations/sinks/utils/writer.py

```diff
@@ -83,20 +86,65 @@ def save_df_as_table(
             if config["options"] is not None and config["options"] != {}:
                 _df_writer = _df_writer.options(**config["options"])
 
-            if config["format"] == "bigquery":
-                _df_writer.option("table", table_name).save()
-            else:
-                _df_writer.saveAsTable(name=table_name)
-                self.spark.sql(
-                    f"ALTER TABLE {table_name} SET TBLPROPERTIES ('product_id' = '{self._context.product_id}')"
+            max_retries = 10
+            for attempt in range(1, max_retries + 2):
+                _log.info(
+                    "Writing records to table: %s, attempt: %s", table_name, attempt
                 )
-            _log.info("finished writing records to table: %s,", table_name)
-
+                try:
+                    if config["format"] == "bigquery":
+                        _df_writer.option("table", table_name).save()
+                    else:
+                        _df_writer.saveAsTable(name=table_name)
+                        if not stats_table:
+                            # Fetch table properties
+                            table_properties = self.spark.sql(
+                                f"SHOW TBLPROPERTIES {table_name}"
+                            ).collect()
+                            table_properties_dict = {
+                                row["key"]: row["value"] for row in table_properties
+                            }
+
+                            # Set product_id in table properties
+                            if (
+                                table_properties_dict.get("product_id") is None
+                                or table_properties_dict.get("product_id")
+                                != self._context.product_id
+                            ):
+                                _log.info(
+                                    "product_id is not set for table %s in tableproperties, setting it now",
+                                    table_name,
+                                )
+                                self.spark.sql(
+                                    f"ALTER TABLE {table_name} SET TBLPROPERTIES ('product_id' = "
```

Comment:
This the main culprit with error. https://docs.databricks.com/en/delta/custom-metadata.html#store-custom-tags-in-table-properties, I guess suggestion is to not club alter with write operations.

Reply:
Removed this for stats table.